api-resources.md

```diff
@@ -0,0 +1,337 @@
+
+We use HTTP `PUT` requests for idempotent creation of resources. Server responses to `PUT` requests follow [RFC 9110 section 9.3.4](https://httpwg.org/specs/rfc9110.html#rfc.section.9.3.4).
+
+Finally, a quick reminder of some HTTP semantics: `GET` and `PUT` are idempotent (i.e., the same request may be made multiple times without changing state in the server more than once, and the response will be the same each time). `POST` is not idempotent. `GET` requests may not have bodies.
+
+We care a lot about idempotence because we need to account for cases where a client makes a request but never sees the response, and thus needs a way to recover from this state of uncertainty.
+
+### Tasks
+
+TODO: do we want a task resource that we can GET and PUT in the protocol? Does this collide with https://wangshan.github.io/draft-wang-ppm-dap-taskprov/draft-wang-ppm-dap-taskprov.html or https://github.com/divergentdave/draft-dcook-ppm-dap-interop-test-design?
+
+### HPKE configurations
+
+#### Path
+
+`/hpke_config[?task_id=task-id]`
+
+The `task_id` query parameter is optional, as described in https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-4.3.1.
+
+##### Alternatives
+
+An HPKE config has an ID, so we could use a path `hpke_configs/hpke-config-id`. However, a client should never need to get any HPKE config except for the "current" one, so does it make sense to expand the API surface with a path param for HPKE config ID?
+
+`GET /hpke_config` could redirect to `/hpke_config/current-hpke-config-id`, which would make this mostly transparent to clients, assuming they handle 3xx redirects properly.
+
+Instead of providing the task ID as a query parameter, we could also include it in the path, like `/tasks/{task-id}/hpke_config`, on the premise that the task ID is part of the resource's permanent identity. However, it's awkward to solve the global HPKE config case with this formulation. We could include a special task ID "global" so that clients could query `/tasks/global/hpke_config`.
+
+#### Representation
+
+Uses the existing [`struct HpkeConfig`](https://www.ietf.org/archive/id/draft-ietf-ppm-dap-02.html#section-4.3.1-6).
+
+```
+struct {
+  HpkeConfigId id;
+  HpkeKemId kem_id;
+  HpkeKdfId kdf_id;
+  HpkeAeadKdfId aead_id;
+  HpkePublicKey public_key;
+} HpkeConfig;
+```
+
+#### Required HTTP methods
+
+##### GET/HEAD
+
+The response body is a `struct HpkeConfig` representing the `hpke_config`.
+
+### Reports
+
+#### Path
+
+`/tasks/{task-id}/reports/{report-id}`
+
+`report-id` uniquely identifies a report in a task. It is chosen by the client when uploading a report.
+
+#### Representation
+
+```
+struct {
+  Time time;
+  Extension extensions<0..2^16-1>;
+} ReportMetadata;
+
+struct {
+  ReportMetadata metadata;
+  opaque public_share<0..2^32-1>;
+  HpkeCiphertext encrypted_input_shares<1..2^32-1>;
+} Report;
+```
+
+#### Required HTTP methods
+
+##### GET/HEAD
+
+The response body is a `struct Report` representing the report.
```

Comment:
Who needs this functionality, and for what purpose? How do we think about how long this report needs to be kept around?

This might be a bit tricky for Daphne because we may want to shard reports by timestamp.

Reply:
You're right, I was just putting it in because semantically it's obvious, but there's no protocol use case, so we shouldn't specify it (implementations are free to implement this if they find it useful).

- [ ] remove GETters we don't explicitly need in the protocol